### PR TITLE
fix(proxy): rad remotes

### LINKED ISSUE #621: Create with existing rad remote
Currently, we have this logic https://github.com/radicle-dev/radicle-upstream/blob/master/proxy/src/coco/peer.rs#L158-L166 that checks if we have a `rad` remote already in the project we're creating. But it's definitely possible to get in a state where the project is not in the monorepo, but the remote still exists in the `.git/config` of the working copy.

Steps to reproduce:
* Add existing repository
* `rm -r ~/.local/share/radicle`
* Try add the same existing repository

This bars us from adding the project to our monorepo which is not great. It might be better to check if the `RadUrn` of the project we're creating exists instead.

### LINKED ISSUE #780: rad:// for project remotes
We currently have remotes set up with `file://` urls. We should be setting them up as `rad://` urls to be able to piggy-back on the rad remote helper.

### PR DESCRIPTION
**Pull request by FintanH**
Fixes #780 
Fixes #621 

This PR addresses a few things:
* Refactor checkout and create into the coco::project module.
* Creation uses the ref and remote apis from librad
* Setting up remotes is now using the rad:// url scheme so that the remote helper kicks in
* We don't error when the remote exists

### CODE REVIEW

File: proxy/src/coco/config.rs [hunk 1]
```diff
@@ -11,6 +12,9 @@ use librad::peer;
 
 use crate::error;
 
+/// The environment variable that points to where librad data lives.
+pub const RAD_HOME: &str = "RAD_HOME";
```
[the remark this answers is not in the document]
Comment (FintanH) [reply proxy/src/coco/config.rs:16]: It's used below when we specify a `Paths::from_root`, basically saying that the path we're using is `RAD_HOME`

File: proxy/src/coco/project.rs [hunk 2]
```diff
@@ -1,247 +1,67 @@
...
-    /// Create a new project where the final directory path is `<path>/<name>`.
-    New {
-        /// The name of the project.
-        name: String,
-        /// The directory where we create the project.
-        path: Path,
-    },
+/// Specify how to create the git credential helper argument for a [`Checkout`]
+enum Credential {
+    /// Plain-text password. You've been warned!
```
[the remark this answers is not in the document]
Comment (FintanH) [reply proxy/src/coco/project.rs:18]: If my checkout MVP checks out then this can go :)

File: proxy/src/coco/project/checkout.rs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            .arg("rad")
+            .arg("--branch")
+            .arg(self.project.default_branch())
+            .arg(LocalUrl::from_urn(self.project.urn(), peer_id).to_string())
+            .arg(project_path.as_os_str())
+            .env("PATH", &bin_path)
+            .envs(std::env::vars().filter(|(key, _)| key.starts_with("GIT_TRACE")))
+            .spawn()?;
+
+        // TODO: Capture the error if any and respond
```
[the remark this answers is not in the document]
Comment (FintanH) [reply proxy/src/coco/project/checkout.rs:111]: Same here: https://github.com/radicle-dev/radicle-upstream/pull/781#discussion_r471404946

**Comment by FintanH** [comment]
>  when I create a new project and add it via Upstream, I can't push new commits from the project directory to radicle:

We need https://github.com/radicle-dev/radicle-link/pull/285 to make this work properly :)

**Comment by xla** [comment]
> We need [radicle-dev/radicle-link#285](https://github.com/radicle-dev/radicle-link/pull/285) to make this work properly :)

@FintanH is this a blocker for this PR, or would you capture it in a follow-up?

**Comment by FintanH** [comment]
> @FintanH is this a blocker for this PR, or would you capture it in a follow-up?

@xla We can't push from the repositories so it's a breaking change. Same with the checkout @rudolfs pointed out. I'd like to get these fixed

**Comment by rudolfs** [comment]
Almost there Finto!
<details>
<summary>❌ When I try to push from a new repo that I just set up, the current branch isn't tracking rad:</summary>

```
radicle-tests/new » git push rad

fatal: The current branch master has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream rad master
```
</details>

<details>
<summary>✅ However, when I create a project from an existing repo, everything works as expected:</summary>

```
radicle-tests/platinum » git push rad
Password for 'rad://radicle@hwd1yrenoptpitcaznpf3hb5cgpcn5n6m1b5bncmpjh5he5i7bo73x5jp6r.git':
Enumerating objects: 5, done.
Counting objects: 100% (5/5), done.
Delta compression using up to 8 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 928 bytes | 928.00 KiB/s, done.
Total 3 (delta 1), reused 0 (delta 0)
To rad://hwd1yrenoptpitcaznpf3hb5cgpcn5n6m1b5bncmpjh5he5i7bo73x5jp6r.git
   223aaf8..3f0fe24  master -> master
```
</details>

<details>
<summary>❌ A checked out working directory also doesn't have branch <-> remote mappings:</summary>

```
radicle-checkout/radicle-upstream » git pull rad
Password for 'rad://radicle@hwd1yrefmsc3yn3p3npk7iahxr3iqr1j7g49x94kgxh9ego18pqzrsbosky.git':
You asked to pull from the remote 'rad', but did not specify
a branch. Because this is not the default configured remote
for your current branch, you must specify a branch on the command line.
```
</details>

<details>
<summary>❌ When creating a new project from an existing repository with a rad remote, we get validation errors in the form:</summary>

  - seems like more work is required to fix #621
  - "create project" button should be disabled if there are errors in the form

Screenshot 2020-08-19 at 14 20 53
</details>

**Comment by FintanH** [comment]
> When creating a new project from an existing repository with a rad remote, we get validation errors in the form:

Turns out we had some old logic looking for `rad` remotes and saying its managed by Radicle already. Removing it :)

**Comment by rudolfs** [comment]
Found another edge case, possibly related to: https://github.com/radicle-dev/radicle-upstream/issues/774.

1. `rm -rf /Users/rudolfs/work/radicle-tests/*`
2. `yarn nuke`
3. `yarn start` -> create identity: rudolfs
4. `new project` -> name: new -> `start with a new repo` -> path to folder: /Users/rudolfs/work/radicle-tests -> `create project`
5. quit app
6. `yarn nuke`
7. `yarn start` -> create identity: rudolfs
8. `new project` -> `continue with existing repo` -> path to folder: /Users/rudolfs/work/radicle-tests/new -> `create project`

Proxy errors:
```
 INFO  tracing::span           > <- Storage::create_repo
 INFO  tracing::span           > -- Storage::create_repo
 INFO  librad::git::storage    > creating symbolic link: refs/namespaces/hwd1yren6k1n9mx3o6ec1quo8x5d9bxxopcq95pdcy45ezso14m9k11rjxo/refs/rad/self -> refs/namespaces/hwd1yredbsi9sgn5e661h7kcfm8dhadf88i8nh4emy37r4pge5nqsfdurrc/refs/rad/id
 INFO  librad::git::types      > creating symbolic ref refs/namespaces/hwd1yren6k1n9mx3o6ec1quo8x5d9bxxopcq95pdcy45ezso14m9k11rjxo/refs/rad/self -> refs/namespaces/hwd1yredbsi9sgn5e661h7kcfm8dhadf88i8nh4emy37r4pge5nqsfdurrc/refs/rad/id
 ERROR proxy::http::error      > Rejection([MethodNotAllowed, Git(Git(Error { code: -3, klass: 4, message: "reference \'refs/namespaces/hwd1yren6k1n9mx3o6ec1quo8x5d9bxxopcq95pdcy45ezso14m9k11rjxo/refs/heads/master\' not found" })), MethodNotAllowed])
 INFO  proxy::http             > "POST /v1/projects HTTP/1.1" 400 164.294724ms
```

UI:

Screenshot 2020-08-20 at 10 37 14
Screenshot 2020-08-20 at 10 37 18

**Review by rudolfs — APPROVED** [review]
LGTM 🚀

File: proxy/src/coco/config.rs [hunk 4]
```diff
@@ -57,6 +64,7 @@ pub fn default(
     key: keys::SecretKey,
     path: impl AsRef<std::path::Path>,
 ) -> Result<net::peer::PeerConfig<Disco, keys::SecretKey>, error::Error> {
+    env::set_var(RAD_HOME, path.as_ref());
```
Comment (cloudhead) [inline proxy/src/coco/config.rs:67]: Note that this will only set the env of this process and its children. Is that the intention?

File: proxy/src/coco/project/create.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+        // Test if the repo has setup rad remote.
+        match repo.find_remote(config::RAD_REMOTE) {
+            Ok(mut remote) => {
+                // Send a warning if the remote urls don't match
+                if let Some(remote_url) = remote.url() {
+                    if remote_url == url.to_string() {
+                        Self::push_default(&mut remote, &self.default_branch)?;
+                    } else {
+                        log::warn!("Remote URL Mismatch: '{} /= '{}'", remote_url, url);
```
Comment (cloudhead) [inline proxy/src/coco/project/create.rs:179]:
```suggestion
                        log::warn!("Remote URL Mismatch: '{}' /= '{}'", remote_url, url);
```

File: proxy/src/coco/config.rs (hunk 4 above)
Comment (FintanH) [reply proxy/src/coco/config.rs:67]: Actually, I think this is useless to me. I thought we needed it for the rad remote helper.

Comment (cloudhead) [reply proxy/src/coco/config.rs:67]: That would make sense, if we're spawning it as a child

**Review by cloudhead — APPROVED** [review]
